Step 3 of dropmixnuc refactor

brigaudet opened the pull request:
Step 3 of ndrop.F90/dropmixnuc refactor:

1)  Create new subroutine update_from_explmix for explicit droplet / aerosol mixing substep procedure.
2)  Move vertical level loop within update_from_newcld
3)  Use only 1-column versions of loadaer, get_aer_mmr_sum, get_aer_num; remove multi-column versions
4)  Move erf testing to ndrop_init subroutine
5)  Add intent labels, parameter declarations; combine / remove code

Commit:
Add in, out, inout notation; remove more unneeded code

Commit:
Move kk loop through levels inside update_from_newcld

Commit:
Create new subroutine update_from_explmix to handle mixing substepping section of dropmixnuc.

Commit:
Remove unused code after last refactoring step.

Commit:
Modify ccncalc to utilize 1-col subroutines

Commit:
Remove non-1-col versions of subroutines since no longer used.

Commit:
Revert ccncalc to non-1col function to restore BFB.

Commit:
Combine code for qcldbrn

Commit:
Adding parameters and some restructure

Commit:
Fixed bug.

jeff-cohere commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -110,9 +110,23 @@ subroutine ndrop_init
    !-------------------------------------------------------------------------------
 
+   arg = 1.0_r8
+   if (abs(0.8427_r8 - erf(arg))/0.8427_r8 > 0.001_r8) then
```
what is the significance of `0.8427`? Can we create a `parameter` for this?

jeff-cohere replied:
Is this subroutine actually testing the error function at runtime?

brigaudet replied:
Here is a link to Balwinder's notes:
https://eagles-project.atlassian.net/wiki/spaces/Computation/pages/edit-v2/1623228417
If the link doesn't work -- I think this is related to the fact that some fortran compilers don't provide Fortran intrinsic erf functions, in which case cime might provide its own error function.  This is to ensure that there is an erf function defined that gives the answers the erf functions in the code are expecting.  It was previously in dropmixnuc but I just moved it to ndrop_init.

To answer the other question -- 0.8427 is just erf(1.0), code is just testing whether this is in fact what the erf function produces.  It could certainly be converted to a parameter.

jeff-cohere replied:
Thanks! C and C++ both define `erf` in their standard libraries, so we can probably skip this exercise in the port.

singhbalwinder replied:
This part of the code was part of the runtime but Brian moved it to the init routine. This was just a sanity check, we can skip this for the C++ port (like @jeff-cohere mentioned)

singhbalwinder commented:
@brigaudet : Are you still working on this PR? I see some comments with your name and commented-out code.

Commit:
Remove extraneous comments.

brigaudet commented:
> @brigaudet : Are you still working on this PR? I see some comments with your name and commented-out code.

@singhbalwinder Sorry, I forgot to remove those.  I've just done so.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -489,25 +454,20 @@ subroutine dropmixnuc( &
    wtke(:,:)    = 0._r8
    factnum(:,:,:) = 0._r8
 
+ !NOTE FOR C++ PORT: Get the cloud borne MMRs from AD in variable qcldbrn, do not port the code before END NOTE
```
Is this comment still relevant? I think the c++ folks should be creating a `qqcw` like data structure to store the cloud-borne aerosols mmr. `qcldbrn` will be populated in the following loops.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -524,10 +484,10 @@ subroutine dropmixnuc( &
 
       nsav = 1
-      nnew = 2
+
```
Make `nsav` and `nnew` parameters as they stay the same for the entire simulation.

brigaudet replied:
Actually, these are not constants.  Subroutine 'update_from_explmix' (as did the previous code) stores two time levels of aerosol fields, denoted by a length-two array dimension whose indices are 'nsav' and 'nnew'.   For each subloop in update_from_explmix, new values are put into the nnew element, while the old values are sent to the nsav element.  However, this is not done by physically copying the arrays, but by switching the values of 'nsav' and 'nnew' for each subloop.  This is also why these values are passed as arguments into 'update_from_explmix' -- because there did not seem to be a straightforward way of determining what the values of 'nsav' and 'nnew'  would be once all subloops are completed.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -914,10 +645,9 @@ subroutine get_activate_frac(  &
 
```
`wmax` can be declared as a parameter

brigaudet replied:
OK, done.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -914,10 +645,9 @@ subroutine get_activate_frac(  &
-            state_q_kload(:), &
-            imode, nspec_amode(imode), cs_kload, phase, na, va, &
-            hy )
+         call loadaer_1col( state_q_kload(:), &  ! in
```
I think it makes more sense to take the `imode` do-loop inside `loadaer_1col1` and change the name of `loadaer_1col` to just `loadaer`

brigaudet replied:
I've gone ahead and done this, as well as the other requested changes, such as creating parameters and renaming the relevant subroutines.  All of the longer term tests PASS.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -1206,22 +951,254 @@ subroutine update_from_cldn_profile(cldn_col_in,dtinv,wtke_col_in,zs,dz,temp_col
+   real(r8), intent(inout) :: qcld(:)  ! cloud droplet number mixing ratio [#/kg]
+   real(r8), intent(inout) :: raercol(:,:,:)    ! single column of saved aerosol mass, number mixing ratios [#/kg or kg/kg]
+   real(r8), intent(inout) :: raercol_cw(:,:,:) ! same as raercol but for cloud-borne phase [#/kg or kg/kg]
+   integer, intent(inout) :: nnew, nsav   ! indices for old, new time levels in substepping
```
You can declare both nsav and nnew as parameters at the top so that we don't need to send them as arguments.

brigaudet replied:
See my discussion about nsav and nnew above.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -1522,11 +1499,12 @@ subroutine ccncalc(state_q, tair, qcldbrn, qcldbrn_num, ncol, cs, ccn)
-            imode, nspec_amode(imode), cs, phase, naerosol, vaerosol, &
-            hygro, &
-            qcldbrn(:,:,kk,imode), qcldbrn_num(:,kk,imode) )  ! optional in
+         do icol = 1, ncol
```
In the line "if(na(1) < 1.e-20_r8)return" above, please declare 1.e-20_r8 as a parameter. Call it "small_val" or something similar. Also, please make "1.e-39_r8" a parameter as well.

The following can also be parameters:
```
percent_to_fraction = 0.01_r8
   per_m3_to_per_cm3 = 1.e-6_r8
   smcoefcoef=2._r8/sqrt(27._r8)
```

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -1522,11 +1499,12 @@ subroutine ccncalc(state_q, tair, qcldbrn, qcldbrn_num, ncol, cs, ccn)
-            hygro, &
-            qcldbrn(:,:,kk,imode), qcldbrn_num(:,kk,imode) )  ! optional in
+         do icol = 1, ncol
+            call loadaer_1col( state_q(icol,kk,:), &  ! in
```
Where you have calculations like "3._r8/(4._r8*pi*exp45logsig(imode))", please make a note to the c++ team that common utility functions (that exist in the C++ code) might be used here. This calculation is used for computing particle size and we are trying to do this calculation in one place.

singhbalwinder commented on components/eam/src/physics/cam/ndrop.F90:
```diff
@@ -1632,92 +1609,9 @@ subroutine loadaer_1col( &
 end subroutine loadaer_1col
```
There is a comment about "loadaer_1cell" subroutine above, please fix it as we don't have any "loadaer_1cell" routine now.

We can rename `get_aer_num_1col` and `get_aer_mmr_sum_1col` by removing `_1col`.

singhbalwinder reviewed (changes requested):
Nice work! The code is easy to navigate and understand. I left some minor comments. Let me know if you have any questions.

Also, the code is not indented properly. Do you have automatic indentation in your editor? If not, I can fix the indentation once you are done with the changes.

Commit:
Create more parameter declarations for constants.

Commit:
Create even more parameters from constants, organize variable declarations and remove unneeded ones.

Commit:
Move icol loop within loadaer, modify other subroutines accordingly
Change order of indices in qcldbrn
Additional parameter specification and reformatting

singhbalwinder committed:
code indentation

singhbalwinder commented:
Thanks for addressing my comments! The code looks great! I have indented the entire file in my last commit. It should not impact the code but I am still being cautious and running tests. I will merge it after my tests pass.

singhbalwinder commented:
Here are the test results:
```
$ /compyfs/sing201/e3sm_scratch//ndrop_all/cs.status.comp_ndrop_all_mam4_org_v2_baselines -s
comp_ndrop_all_mam4_org_v2_baselines: 9 tests
  PASS ERP_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS ERS.ne4pg2_oQU480.F2010.compy_intel
  PASS PEM_P64x1.ne4pg2_oQU480.F2010.compy_intel
  PASS PET.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P1x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D_Ln5_P32x1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_D.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS_Ly1.ne4pg2_oQU480.F2010.compy_intel
  PASS SMS.ne30pg2_EC30to60E2r2.F2010.compy_intel
```
